### PR TITLE
Add a unit test for backend create()

### PR DESCRIPTION
**Pull request by asalkeld**

### CODE REVIEW

File: pkg/backend/openshiftcluster/create.go [hunk 1]
```diff
@@ -161,7 +160,11 @@ func (m *Manager) Create(ctx context.Context) error {
 		return err
 	}
 
-	return install.NewInstaller(m.log, m.env, m.db, m.fpAuthorizer, r.SubscriptionID).Install(ctx, m.doc, installConfig, platformCreds)
+	if m.installer == nil {
+		m.installer = install.NewInstaller(m.log, m.env, m.db, m.fpAuthorizer, r.SubscriptionID)
```
Comment (m1kola) [inline pkg/backend/openshiftcluster/create.go:164]: It looks like this `if` only exists for tests. Does it make sense to move `install.NewInstaller()` to the calling code? For example, create it somewhere in `openShiftClusterBackend#handle` and pass into `openshiftcluster.NewManager`?

File: pkg/backend/openshiftcluster/create_test.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	defer gmc.Finish()
+	logger := logrus.NewEntry(logrus.StandardLogger())
+
+	doc := &api.OpenShiftClusterDocument{
+		Key:              "foo",
+		OpenShiftCluster: mocks.MockOpenShiftCluster(),
+	}
+
+	mDB := mock_database.NewMockOpenShiftClusters(gmc)
+	mDB.EXPECT().Patch(gomock.Any(), gomock.Any()).Return(doc, nil)
```
Comment (m1kola) [inline pkg/backend/openshiftcluster/create_test.go:32]: `Manager#Create` seem to mainly consist of calls of other functions which ideally should be tested separately (like `subnet.Split` and db's `Patch`, `validation.ValidateInstallConfig` from installer, etc).

Hovever there are few pieces that potentially need testing in `Manager#Create`:
* Document patch callback where we populate properties `SSHKey`, `StorageSuffix`, `DomainName` if they are not present.
* Constructed `installConfig` and `platformCreds`

But we do not cover them here: we replace them with `gomock.Any()`

File: pkg/util/mocks/openshiftcluster.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
+package mocks
```
Comment (m1kola) [inline pkg/util/mocks/openshiftcluster.go:1]: I would prefer `pkg/util/mocks` to only contain generated mocks to be able to cleanup easily: just remove who dir and regenerate everything.

**Review by m1kola — CHANGES_REQUESTED** [review]

File: pkg/backend/openshiftcluster/create.go (hunk 1 above)
Comment (jim-minter) [reply pkg/backend/openshiftcluster/create.go:164]: +1 - might making it a member of openshiftcluster.Manager and initialising it in NewManager be good enough?

File: pkg/util/mocks/openshiftcluster.go (hunk 3 above)
Comment (jim-minter) [reply pkg/util/mocks/openshiftcluster.go:1]: somewhere under the newly created `test/` for this and the stuff below probably makes sense.

File: pkg/util/mocks/tls.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
+package mocks
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+)
+
+var (
+	DummyPrivateKey  *rsa.PrivateKey
+	DummyCertificate *x509.Certificate
```
Comment (jim-minter) [inline pkg/util/mocks/tls.go:10]: not used

**Comment by jim-minter** [comment]
needs rebase.  A lot of the foundational stuff makes sense here; I feel like a(nother) conversation about testing strategy would be good.  For example I wonder whether it makes sense to try and test from `backend.Run()` as far as `installer.Install()`.  I'm not sure.

**Comment by m1kola** [comment]
@asalkeld @jim-minter I think, it is possible to cover create, update and delete with more or less useful unit tests, but I'm not sure if unit tests are good fit here. I expect unit tests to be very fragile in this part of the project and require a lot of maintanance due to the fact that there are at lot of things to mock.

Maybe we should cover underlying code (`database`, `subnet` packages, etc) with unit tests and start looking into e2e tests?

**Comment by asalkeld** [comment]
> @asalkeld @jim-minter I think, it is possible to cover create, update and delete with more or less useful unit tests, but I'm not sure if unit tests are good fit here. I expect unit tests to be very fragile in this part of the project and require a lot of maintanance due to the fact that there are at lot of things to mock.
> 
> Maybe we should cover underlying code (`database`, `subnet` packages, etc) with unit tests and start looking into e2e tests?

So, I partly agree...
IMHO we should unit test as much as possible. BUT we need to structure the code to make this sensible and useful. I like the idea of testing in layers (based on interfaces), but it also seems be dependent on each layer doing enough (to make it worth while), and not too much to make it a horrible mocking nightmare. I created this PR (I intend to continue work on it, by adding one test for each layer) to see what the the tests in each layer looked like. I am not in favor of punting unit tests in some areas, but for re-structuring until the tests are both useful and maintainable.

**Pull request closed without merging**